Added `interleaveAll` operation

ahmadsalim opened the pull request:
Additionally added tests for both `interleave` and `interleaveAll`.

ahmadsalim commented:
The `interleaveAll` operation basically keeps emitting values from the longer stream when the shorter stream is halted, instead of cutting off like `interleave` does. I hope it is useful, also in conjunction with the new redesign.

pchiusano commented:
@ahmadsalim would you like to try implementing this for the new representation (in `fs2` package)? I probably won't merge this as is, just because we're directing most of our efforts toward the FS2 work and wouldn't likely do another release off the old design unless it's like a serious bugfix and the new release is still a ways off.

ahmadsalim commented:
@pchiusano Sure

ahmadsalim commented:
@pchiusano OK, I have updated the implementation for `fs2` as well. Do you want me to remove the old one in `scalaz.stream`?

pchiusano commented:
Yeah, you can remove it.

I have one more request. I didn't want to bring it up earlier to discourage you, but I'd like to preserve the chunkiness of the input streams as much as possible. This will be much more efficient. So use `receive` rather than `receive1` to obtain chunks, then interleave the chunks. The code will be more complicated, since you'll need to deal with the case where the chunks have different sizes.

ahmadsalim commented:
OK, I will take a look.

pchiusano commented on src/main/scala/fs2/StreamDerived.scala:
```diff
@@ -57,6 +57,9 @@ private[fs2] trait StreamDerived { self: fs2.Stream.type =>
   def eval_[F[_],A](fa: F[A]): Stream[F,Nothing] =
     flatMap(eval(fa)) { _ => empty }
 
+  def infinite[F[_],W](ws: W*): Stream[F, W] =
```
Pls call this `constant` to be consistent with existing signature in old code:

```Scala
  /**
   * The infinite `Process`, always emits `a`.
   * If for performance reasons it is good to emit `a` in chunks,
   * specify size of chunk by `chunkSize` parameter
   */
  def constant[A](a: A, chunkSize: Int = 1): Process0[A]
```

Also, just have it take a single `W`. If you want to repeat a sequence over and over, just do `emits(ws).repeat`.

pchiusano commented on src/main/scala/fs2/Tee.scala:
```diff
@@ -10,6 +12,115 @@ object tee {
+          P.output(co)
+        }
+        def contLeft(h: Handle[F,I]): Pull[F,O,Nothing] = {
+          P.receive((s: Step[Chunk[I], Handle[F, I]]) => s match {
```
just do `h.receive { case c #: h => ... }`.

pchiusano commented on src/main/scala/fs2/Tee.scala:
```diff
@@ -10,6 +12,115 @@ object tee {
+      _.pull2(_)(goB)
+  }
+
+  def zipAllWith[F[_],I,I2,O](pad1: I, pad2: I2)(f: (I, I2) => O): (Stream[F, I], Stream[F, I2]) => Stream[F, O] = {
```
wow this got pretty complicated

pchiusano commented on src/main/scala/fs2/Tee.scala:
```diff
@@ -10,6 +12,115 @@ object tee {
+          P.output(co)
+        }
+        def contRight(h: Handle[F,I2]): Pull[F,O,Nothing] = {
+          P.receive((s: Step[Chunk[I2], Handle[F, I2]]) => s match {
```
`h.receive { case c #: h => ... }`

pchiusano commented on src/test/scala/fs2/TeeSpec.scala:
```diff
@@ -0,0 +1,48 @@
+import org.scalacheck.Prop._
+import org.scalacheck.{Gen, Properties}
+
+object TeeSpec extends Properties("tee") {
```
tests for the zip functions?

pchiusano commented on src/main/scala/fs2/Tee.scala:
```diff
@@ -10,6 +12,115 @@ object tee {
+  def interleave[F[_], O]: (Stream[F,O], Stream[F,O]) => Stream[F,O] =
+    zip(_,_) flatMap { case (i1,i2) => Stream(i1,i2) }
+
+  def interleaveAll[F[_], O]: (Stream[F,O], Stream[F,O]) => Stream[F,O] = { (s1, s2) =>
```
cute implementation!

the output doesn't preserve chunkiness though

pchiusano commented:
Sorry for the delay in responding, I didn't see you'd updated the PR (I get notified if you add a comment, not if you just add a commit to the PR).

I actually have a bunch of things I want to do after this is merged to see if it's possible to clean up, however I think it'll be easier if I just work on this after merging rather than doing a bunch of back and forth.

However, can you just:

* Just rename `infinite` to `constant` and give it an optional default `chunkSize` parameter like the old signature.
* Add some tests for the `zip` functions. I know they get some ambient testing via `interleave` tests.

Then I'll merge.

ahmadsalim commented:
@pchiusano Sure, I will do that. 

I actually wanted to notify you after I had added some tests and cleaned up a bit, but I got preoccupied with a few things at work.

pchiusano commented:
Np, thanks!

On Tue, Dec 8, 2015 at 10:17 AM Ahmad Salim Al-Sibahi <
notifications@github.com> wrote:

> @pchiusano <https://github.com/pchiusano> Sure, I will do that.
>
> I actually wanted to notify you after I had added some tests and cleaned
> up a bit, but I got preoccupied with a few things at work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/functional-streams-for-scala/fs2/pull/496#issuecomment-162911840>
> .
>

ahmadsalim commented:
@pchiusano OK, I tried to clean up a bit as you suggested and renamed `infinite` to `constant` and added tests for the `zip*` functions. Sorry for the complex code, I wanted to clean it up more but fell a bit short on time; I am glad that you will try to simplify it.

pchiusano commented:
Okay cool merging.
